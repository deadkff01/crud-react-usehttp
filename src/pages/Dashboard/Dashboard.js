import React, { useState, useEffect } from "react";
import { useNavigate } from "@reach/router";
import { useHttpInstance } from "services/baseUseHttp";
import Loader from "components/Loader";

const Dashboard = () => {
  const navigate = useNavigate();
  const [items, setItems] = useState([]);
  const { get, del, response, loading, error } = useHttpInstance();

  useEffect(() => {
    loadItems();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  async function loadItems() {
    const itemsResponse = await get("get-items");
    console.log(itemsResponse);
    if (response.ok) setItems(itemsResponse);
  }

  const deleteItem = async (id) => {
    await del(`delete-by-id/${id}`);
    if (response.ok) setItems(items.filter((item) => item.id !== id));
  };

  const goToItem = (id) => {
    navigate(`/item/${id}`);
  };

  if (error) {
    return <div>Error</div>;
  }

  if (loading) {
    return <Loader />;
  }

  return (
    <div className="container mx-auto" data-testid="container">
      <h1 className="mt-5 text-4xl font-bold">Consoles</h1>
      <div className="mb-5 text-right">
        <button
          onClick={() => navigate(`/add-item`)}
          className="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded"
        >
          Add Item
        </button>
      </div>
      <ul className="grid grid-cols-3 gap-10">
        {items.map(({ id, name, price }) => (
          <li
            key={id}
            className="border-solid border-2 border-gray-600 max-w-sm rounded overflow-hidden shadow-lg p-2"
          >
            <h1 className="text-xl font-bold">{name}</h1>
            <h2 className="text-lg font-bold">{price}</h2>
            <button
              onClick={() => goToItem(id)}
              className="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded mr-5"
            >
              Details
            </button>
            <button
              onClick={() => deleteItem(id)}
              className="bg-red-500 hover:bg-red-700 text-white font-bold py-2 px-4 rounded"
            >
              Remove
            </button>
          </li>
        ))}
      </ul>
    </div>
  );
};

export default Dashboard;
