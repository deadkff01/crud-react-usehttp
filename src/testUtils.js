import React from "react";
import {
  Router,
  createHistory,
  createMemorySource,
  LocationProvider,
} from "@reach/router";
import { render } from "@testing-library/react";

export function renderWithRouter(
  ui,
  { route = "/", history = createHistory(createMemorySource(route)) } = {}
) {
  return {
    ...render(<LocationProvider history={history}>{ui}</LocationProvider>),
    history,
  };
}

export function renderWithRouterParams(
  ui,
  { route = "/", history = createHistory(createMemorySource(route)) } = {}
) {
  return {
    ...render(
      <LocationProvider history={history}>
        <Router>{ui}</Router>
      </LocationProvider>
    ),
    history,
  };
}
