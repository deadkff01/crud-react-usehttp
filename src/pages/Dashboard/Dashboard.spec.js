import React from "react";
import {
  cleanup,
  waitForElement,
  fireEvent,
  wait,
  screen,
  act,
} from "@testing-library/react";
import { rest } from "msw";
import { setupServer } from "msw/node";
import { renderWithRouter } from "../../testUtils";
import Dashboard from "./Dashboard";
import itemsMock from "__mocks__/items";
import "../../setupTests";

const server = setupServer();

describe("Dashboard component", () => {
  beforeAll(() => server.listen());

  afterEach(() => {
    jest.useFakeTimers();
    server.resetHandlers();
  });

  afterAll(() => {
    jest.useRealTimers();
    server.close();
  });

  test("renders the component", async () => {
    const { getAllByTestId } = renderWithRouter(<Dashboard />);
    await act(async () => {
      server.use(
        rest.get("http://localhost:8080/get-items", (req, res, ctx) => {
          return res(ctx.status(200), ctx.json(itemsMock));
        })
      );
    });
    const element = await waitForElement(() => getAllByTestId("container"));
    expect(element).toMatchSnapshot();
  });

  test("renders the loader", async () => {
    const { getAllByTestId } = renderWithRouter(<Dashboard />);
    await act(async () => {
      server.use(
        rest.get("http://localhost:8080/get-items", (req, res, ctx) => {
          return res(ctx.status(200), ctx.json(itemsMock));
        })
      );
    });
    const element = await waitForElement(() =>
      getAllByTestId("spinner-loader")
    );
    expect(element).toBeDefined();
  });
});
