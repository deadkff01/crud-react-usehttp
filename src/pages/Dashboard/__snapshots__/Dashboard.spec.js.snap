// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard component renders the component 1`] = `
Array [
  <div
    class="container mx-auto"
    data-testid="container"
  >
    <h1
      class="mt-5 text-4xl font-bold"
    >
      Consoles
    </h1>
    <div
      class="mb-5 text-right"
    >
      <button
        class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded"
      >
        Add Item
      </button>
    </div>
    <ul
      class="grid grid-cols-3 gap-10"
    >
      <li
        class="border-solid border-2 border-gray-600 max-w-sm rounded overflow-hidden shadow-lg p-2"
      >
        <h1
          class="text-xl font-bold"
        >
          Nintendo Switch
        </h1>
        <h2
          class="text-lg font-bold"
        >
          2500
        </h2>
        <button
          class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded mr-5"
        >
          Details
        </button>
        <button
          class="bg-red-500 hover:bg-red-700 text-white font-bold py-2 px-4 rounded"
        >
          Remove
        </button>
      </li>
      <li
        class="border-solid border-2 border-gray-600 max-w-sm rounded overflow-hidden shadow-lg p-2"
      >
        <h1
          class="text-xl font-bold"
        >
          Playstation 4
        </h1>
        <h2
          class="text-lg font-bold"
        >
          1500
        </h2>
        <button
          class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded mr-5"
        >
          Details
        </button>
        <button
          class="bg-red-500 hover:bg-red-700 text-white font-bold py-2 px-4 rounded"
        >
          Remove
        </button>
      </li>
    </ul>
  </div>,
]
`;
